File: LintCode/cpp/Subarray Sum II.cc
class Solution {
public:
    /**
     * @param A an integer array
     * @param start an integer
     * @param end an integer
     * @return the number of possible answer
     */
    int subarraySumII(vector<int>& A, int start, int end) {
        int a = solve(A, [start](int x) {
            return x >= start;
        });
        int b = solve(A, [end](int x) {
            return x > end;
        });
        return a - b;
    }
private:
    int solve(const vector<int>& A, function<int(int)> fun) {
        int n = A.size();
        int sum = A[0];
        int p = -1;
        int q = 0;
        int ans = 0;
        
        while (p < n && q < n) {
            sum -= p >= 0? A[p]: 0;
            while (q < n && !fun(sum)) {
                q++;
                sum += A[q];
            }
            ans += n - q;
            p++;
        }
        return ans;
    }
};

File: LintCode/python/Subarray Sum II.py
class Solution:
    # @param {int[]} A an integer array
    # @param {int} start an integer
    # @param {int} end an integer
    # @return {int} the number of possible answer
    def subarraySumII(self, A, start, end):
        A = [0] + A
        n = len(A)
        for i in xrange(1, n):
            A[i] += A[i - 1]
        a = self.solve(A, lambda x: x >= start)
        b = self.solve(A, lambda x: x > end)
        return a - b
    
    def solve(self, A, fun):
        n = len(A)
        ans = 0
        p, q = 0, 1
        while p < n and q < n:
            while q < n and not fun(A[q] - A[p]):
                q += 1
            ans += n - q
            p += 1
        return ans
